File: utilities/RTP/Emissions/join_networklinks_to_emissionrates.R
#
# This R script opens the output network csvs
# and left_joins them to the emissions rates based on county, model year, care/non-care
# where "Speed" is joined on int(CSPD)
#
# PlanBayArea2050 / RTP2021 Asana Task: https://app.asana.com/0/316552326098022/1200007791721297/f

library(dplyr)
library(tidyr)
library(stringr)
library(readxl)

LOOKUP_COUNTY <- data.frame(county2      =c("AL", "CC", "MA", "NA", "SF", "SM", "SC", "SL", "SN"),
                            county_census=c("001","013","041","055","075","081","085","095","097"),
                            county_name  =c("Alameda", "Contra Costa", "Marin", "Napa", "San Francisco", "San Mateo", "Santa Clara", "Solano", "Sonoma"),
                            stringsAsFactors = FALSE)
MODEL_DIRS <- 
  c(IP_2015  ="M:/Application/Model One/RTP2021/IncrementalProgress/2015_TM152_IPA_16",
    NP_2050  ="M:/Application/Model One/RTP2021/Blueprint/2050_TM152_FBP_NoProject_22",
    FBP_2050 ="M:/Application/Model One/RTP2021/Blueprint/2050_TM152_FBP_PlusCrossing_21",
    Alt1_2050="M:/Application/Model One/RTP2021/Blueprint/2050_TM152_EIR_Alt1_03",
    Alt2_2050="M:/Application/Model One/RTP2021/Blueprint/2050_TM152_EIR_Alt2_02")

# these are the shapefile exports of the inputs used in the above directories, corresponded to CARE and counties (link_to_COUNTY_CARE.csv)
NETWORK_CARE_DIRS <-
  c(IP_2015  ="M:/Application/Model One/RTP2021/IncrementalProgress/Networks/network_2015_03/shapefile",
    NP_2050  ="M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_53/net_2050_Baseline/shapefile",
    FBP_2050 ="M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_53/net_2050_Blueprint Plus Crossing/shapefile",
    Alt1_2050="M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_56/net_2050_Alt1/shapefile",
    Alt2_2050="M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_58/net_2050_Alt2/shapefile")

# all files are expected to be here (note R's preference for slashes)
if (Sys.getenv("USERNAME") == "lzorn") {
  BASE_DIR <- "C:/Users/lzorn/Box/CARE Communities/CARE-Data-PBA50"
} else {
  BASE_DIR <- "~/Desktop/CARE-Data_R"
}

# for avgload5period_vehclasses.csv
index_cols   <- c("a","b")
special_cols <- c("distance","lanes","ft","at","tollclass","ffs")

emissions_rates <- list()
# use Tableau wildcard union -- no need to put these in a single file
# loop through each network file
for (network in c("IP_2015","NP_2050","FBP_2050","Alt1_2050","Alt2_2050")) {
  model_dir  <- basename(MODEL_DIRS[network])
  model_year <- substr(model_dir,0,4)
  print(paste("Processing",network," for year",model_year,":",model_dir))
  
  network_file <- file.path(MODEL_DIRS[network],"OUTPUT","avgload5period_vehclasses.csv")
  network_df   <- read.table(network_file, header=TRUE, sep=",", stringsAsFactors=TRUE)
  print(paste(" Read",nrow(network_df),"rows from",network_file))
  print(str(network_df))
  
  # keep subset of columns
  network_df <- select(network_df, c(index_cols,special_cols, 
                                     starts_with("cspd"),                             # congested speed
                                     intersect(starts_with("vol"), ends_with("tot"))  # total volume
                                    ))
  print(head(network_df))
  
  # move timeperiod to rows for cspd
  cspd_df    <- select(network_df, c(index_cols, starts_with("cspd")))
  cspd_tp_df <- pivot_longer(cspd_df, starts_with("cspd"), names_to="time_period", names_prefix="cspd", values_to="cspd")
  # create int version for joining
  cspd_tp_df <- mutate(cspd_tp_df, cspd_int=as.integer(trunc(cspd)))
  print(head(cspd_tp_df))
  
  # move timeperiod to rows for voltot
  vol_df     <- select(network_df, c(index_cols, starts_with("vol")))
  vol_tp_df  <- pivot_longer(vol_df, starts_with("vol"), names_to="time_period", names_pattern="vol(.*)_tot", values_to="voltot")
  print(head(vol_tp_df))
  stopifnot(nrow(vol_tp_df) == nrow(cspd_tp_df))
  
  # put them back together
  network_long_df <- select(network_df, c(index_cols,special_cols))
  network_long_df <- full_join(network_long_df, cspd_tp_df, by=c("a","b"))
  network_long_df <- left_join(network_long_df, vol_tp_df,  by=c("a","b","time_period"))
  # set vmt
  network_long_df <- mutate(network_long_df, vmttot=voltot*distance)
  stopifnot(nrow(network_long_df) == nrow(cspd_tp_df))
  
  # read mapping to CARE
  care_file <- file.path(NETWORK_CARE_DIRS[network],"link_to_COUNTY_CARE.csv")
  care_df   <- read.table(care_file, header=TRUE, sep=",")
  print(paste("Read", nrow(care_df), "lines from",care_file,"; head:"))
  print(head(care_df))

  # combine with network_long
  network_long_care_df <- left_join(network_long_df, 
                                    select(care_df, A, B, COUNTYCARE, linkCC_share),
                                    by=c("a"="A", "b"="B"))
  # set county,CARE from COUNTYCARRER
  network_long_care_df <- mutate(network_long_care_df, 
                                 county_census=substr(COUNTYCARE,0,3),
                                 CARE=ifelse(str_length(COUNTYCARE)>3,1,0))
  
  # filter out dummy links
  network_long_care_df <- filter(network_long_care_df, ft != 6)
  print(paste("network_long_care_df has", nrow(network_long_care_df),"rows"))
  
  # read exhaust-based emissions
  if (is.null(emissions_rates[[paste0("AL-",model_year)]])) {
    emissions_file <- file.path(BASE_DIR, "PBA50_COC_ER_Lookups", paste("Year",model_year,"MSAT Emission Rates with E2021 (PBA2050).xlsx"))
    for (county2 in LOOKUP_COUNTY$county2) {
      sheet_name   <- paste0(county2,"-",model_year,".csv")
      emissions_df <- read_excel(emissions_file, sheet=sheet_name)
      # rename "-" and "." to "_"
      emissions_df <- emissions_df %>% setNames(gsub("\\-","_",names(.))) %>% setNames(gsub("\\.","_",names(.)))
      
      index_name   <- paste0(county2,"-",model_year)
      emissions_rates[[index_name]] <- emissions_df
    }
  }
  
  # read non-exhaust-based emission rates
  index_name <- paste0("non-exhaust-",model_year)
  if (is.null(emissions_rates[[index_name]])) {
    emissions_file <- file.path(BASE_DIR, "PBA50_COC_ER_Lookups", "PM2.5 Non-Exhaust ERs (w E2021 & Mar 2021 road dust update).xlsx")
    emissions_df   <- read_excel(emissions_file, sheet=model_year, skip=1)
    emissions_df   <- emissions_df %>% setNames(gsub(" ","_",names(.)))
    
    emissions_rates[[index_name]] <- emissions_df
  }
  
  # finally, join to emissions
  network_long_care_df <- mutate(network_long_care_df, 
                                 EO_PM2_5    =0.0,
                                 EO_Benzene  =0.0,
                                 EO_Butadiene=0.0,
                                 EO_DieselPM =0.0,
                                 ##### non-exhaust #####
                                 Tire_Wear          =0.0,
                                 Brake_Wear         =0.0,
                                 Entrained_Road_Dust=0.0)
  for (row in 1:nrow(LOOKUP_COUNTY)) {
    this_county2       <- LOOKUP_COUNTY[row, "county2"]
    this_county_census <- LOOKUP_COUNTY[row, "county_census"]
    this_county_name   <- LOOKUP_COUNTY[row, "county_name"]
    
    index_name    <- paste0(this_county2,"-",model_year)
    # print(head(emissions_rates[[index_name]]))
    
    non_exhaust_rates <- filter(emissions_rates[[paste0('non-exhaust-',model_year)]], County==this_county_name)
    
    # filter for this county
    this_co_df   <- filter(network_long_care_df, county_census == this_county_census)
    other_co_df  <- filter(network_long_care_df, (county_census != this_county_census) | is.na(county_census))
    stopifnot(nrow(this_co_df) + nrow(other_co_df) == nrow(network_long_care_df))
    
    # join based on congested speed int
    this_co_df <- left_join(this_co_df, emissions_rates[[index_name]], 
                            by=c("cspd_int"="Speed"))
    this_co_df <- mutate(this_co_df, 
                         EO_PM2_5    =ifelse(CARE==1, C_EO_PM2_5,     NC_EO_PM2_5),
                         EO_Benzene  =ifelse(CARE==1, C_EO_Benzene,   NC_EO_Benzene),
                         EO_Butadiene=ifelse(CARE==1, C_EO_Butadiene, NC_EO_Butadiene),
                         EO_DieselPM =ifelse(CARE==1, C_EO_DieselPM,  NC_EO_DieselPM),
                         # non-exhaust
                         Tire_Wear          =ifelse(CARE==1, non_exhaust_rates[['C_Tire_Wear'          ]], non_exhaust_rates[['NC_Tire_Wear'          ]]),
                         Brake_Wear         =ifelse(CARE==1, non_exhaust_rates[['C_Brake_Wear'         ]], non_exhaust_rates[['NC_Brake_Wear'         ]]),
                         Entrained_Road_Dust=ifelse(CARE==1, non_exhaust_rates[['C_Entrained_Road_Dust']], non_exhaust_rates[['NC_Entrained_Road_Dust']]))
    this_co_df <- select(this_co_df, 
                         -C_EO_PM2_5,     -NC_EO_PM2_5,
                         -C_EO_Benzene,   -NC_EO_Benzene,
                         -C_EO_Butadiene, -NC_EO_Butadiene,
                         -C_EO_DieselPM,  -NC_EO_DieselPM)
    network_long_care_df <- rbind(this_co_df, other_co_df)
  }

  # write the result
  output_fullpath <- file.path(BASE_DIR, "links_CARE_EMFAC2021", paste0("links_CARE_",model_dir,".csv"))
  write.table(network_long_care_df, output_fullpath, sep=",", row.names=FALSE)
  print(paste("Wrote",nrow(network_long_care_df),"rows to",output_fullpath))
}

File: utilities/RTP/Emissions/join_networklinks_to_emissionrates.py
USAGE = """
    Joins the output network csvs with 1) crosswalks of link-CARE Communities (RTP2021/PBA50) and link-Overburdened Communities (RTP2025/PBA50+),
    and 2) the emission rates by county, model year, and CARE/non-CARE or Overburdened/non-Overburdened; emission rates for pollutants other than PM2.5
    are also based on speed, which is joined to int(CSPD) in the network output.

    PlanBayArea2050 / RTP2021 Asana Task: https://app.asana.com/0/316552326098022/1200007791721297/f
    PlanBayArea2050 / RTP2021 Amd1 Asana Task: https://app.asana.com/0/1201730518396783/1208059626391587/f
    PlanBayArea2050+ / RTP2025 Asana Task: https://app.asana.com/1/11860278793487/project/1159042832247728/task/1209490522594295?focus=true

"""

import argparse, os
import pandas as pd
import numpy as np

# Lookup table for counties
LOOKUP_COUNTY = pd.DataFrame({
    "county2": ["AL", "CC", "MA", "NA", "SF", "SM", "SC", "SL", "SN"],
    "county_census": ["001", "013", "041", "055", "075", "081", "085", "095", "097"],
    "county_name": ["Alameda", "Contra Costa", "Marin", "Napa", "San Francisco", "San Mateo", "Santa Clara", "Solano", "Sonoma"]
})

if __name__ == '__main__':

    parser = argparse.ArgumentParser(description = USAGE,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument('rtp', type=str, choices=['RTP2021','RTP2025'], help='Specify RTP')
    my_args = parser.parse_args()


    if my_args.rtp == 'RTP2021':
        MODEL_DIRS = {
            # "IP_2015": "M:/Application/Model One/RTP2021/IncrementalProgress/2015_TM152_IPA_17",
            "NP_2050": "M:/Application/Model One/RTP2021/Blueprint/2050_TM152_FBP_NoProject_24",
            # "FBP_2050": "M:/Application/Model One/RTP2021/Blueprint/2050_TM152_FBP_PlusCrossing_24",
            "FBP_Amd1_2050": "M:/Application/Model One/RTP2021/Blueprint/2050_TM152_FBP_PlusCrossing_24_Amd1",
            # "Alt1_2050": "M:/Application/Model One/RTP2021/Blueprint/2050_TM152_EIR_Alt1_06",
            # "Alt2_2050": "M:/Application/Model One/RTP2021/Blueprint/2050_TM152_EIR_Alt2_05"
        }
    elif my_args.rtp == 'RTP2025':
        MODEL_DIRS = {
            "IP_2023": "M:/Application/Model One/RTP2025/IncrementalProgress/2023_TM161_IPA_34",
            "NP_2050": "M:/Application/Model One/RTP2025/Blueprint/2050_TM161_FBP_NoProject_02",
            "FBP_2050": "M:/Application/Model One/RTP2025/Blueprint/2050_TM161_FBP_Plan_06",
            # "Alt1_2050": "M:/Application/Model One/RTP2025/Blueprint/2050_TM152_EIR_Alt1_06",
            # "Alt2_2050": "M:/Application/Model One/RTP2025/Blueprint/2050_TM152_EIR_Alt2_05"
        }
    if my_args.rtp == 'RTP2021':
        NETWORK_DIRS = {
            # "IP_2015": "M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_62/net_2015_Baseline/shapefile",
            "NP_2050": "M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_62/net_2050_Baseline/shapefile",
            # "FBP_2050": "M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_64/net_2050_Blueprint/shapefile",
            "FBP_Amd1_2050": "M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_64_Amd1/net_2050_Blueprint/shapefile",
            # "Alt1_2050": "M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_64/net_2050_Alt1/shapefile",
            # "Alt2_2050": "M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_64/net_2050_Alt2/shapefile"
        }
    elif my_args.rtp == 'RTP2025':
        NETWORK_DIRS = {
            "IP_2023": "M:/Application/Model One/RTP2025/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_v31/net_2023_Baseline/shapefile",
            "NP_2050": "M:/Application/Model One/RTP2025/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_v31/net_2050_Baseline/shapefile",
            "FBP_2050": "M:/Application/Model One/RTP2025/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_v31/net_2050_Blueprint/shapefile",
            # "Alt1_2050": "M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_64/net_2050_Alt1/shapefile",
            # "Alt2_2050": "M:/Application/Model One/RTP2021/Blueprint/INPUT_DEVELOPMENT/Networks/BlueprintNetworks_64/net_2050_Alt2/shapefile"
        }

    # Set base directory
    if my_args.rtp == 'RTP2021':
        if os.getenv("USERNAME") == "lzorn":
            BASE_DIR = "E:/Box/CARE Communities/CARE-Data-PBA50"
        else:
            BASE_DIR = "M:/Application/PBA50Plus_Data_Processing/OverburdenedCommunities_analysis/PBA50_reproduce"
        EMISSION_RATES_DIR = os.path.join(BASE_DIR, "PBA50_COC_ER_Lookups")
    elif my_args.rtp == 'RTP2025':
        BASE_DIR = "M:/Application/PBA50Plus_Data_Processing/OverburdenedCommunities_analysis"
        EMISSION_RATES_DIR = os.path.join(BASE_DIR, "EmissionRates_Lookups")
    
    # Name of the analysis: "CARE" in RTP2021 and "OverBurdened" in RTP2025
    if my_args.rtp == 'RTP2021':
        analysis_name = "CARE"
    elif my_args.rtp == 'RTP2025':
        analysis_name = "OverBurdened"

    # Index and special columns
    index_cols = ["a", "b"]
    special_cols = ["distance", "lanes", "ft", "at", "tollclass", "ffs"]

    # Initialize emissions rates
    emissions_rates = {}

    # Process network files
    for network in list(NETWORK_DIRS):
        model_dir = os.path.basename(MODEL_DIRS[network])
        model_year = model_dir[:4]
        print(f"Processing {network} for year {model_year}: {model_dir}")

        network_file = os.path.join(MODEL_DIRS[network], "OUTPUT", "avgload5period_vehclasses.csv")
        print(network_file)
        network_df = pd.read_csv(network_file)
        print(f"Read {len(network_df)} rows from {network_file}")

        # Keep subset of columns
        network_df = network_df[index_cols + special_cols + 
                                [col for col in network_df.columns if col.startswith("cspd")] + 
                                [col for col in network_df.columns if col.startswith("vol") and col.endswith("tot")]]
        print('network_df.head()')
        print(network_df.head())

        # Pivot time periods for cspd
        cspd_df = network_df[index_cols + [col for col in network_df.columns if col.startswith("cspd")]]
        cspd_tp_df = cspd_df.melt(id_vars=index_cols, var_name="time_period", value_name="cspd")
        cspd_tp_df['time_period'] = cspd_tp_df['time_period'].apply(lambda x: x.replace("cspd", ""))
        cspd_tp_df["cspd_int"] = cspd_tp_df["cspd"].apply(np.floor).astype(int)
        print(cspd_tp_df.head())

        # Pivot time periods for voltot
        vol_df = network_df[index_cols + [col for col in network_df.columns if col.startswith("vol")]]
        vol_tp_df = vol_df.melt(id_vars=index_cols, var_name="time_period", value_name="voltot")
        vol_tp_df['time_period'] = vol_tp_df['time_period'].apply(lambda x: x.replace("vol", ""))
        vol_tp_df['time_period'] = vol_tp_df['time_period'].apply(lambda x: x.replace("_tot", ""))
        print(vol_tp_df.head())
        assert len(vol_tp_df) == len(cspd_tp_df)

        # Combine dataframes
        network_long_df = network_df[index_cols + special_cols].merge(cspd_tp_df, on=index_cols, how="outer")
        network_long_df = network_long_df.merge(vol_tp_df, on=index_cols + ["time_period"], how="left")
        network_long_df["vmttot"] = network_long_df["voltot"] * network_long_df["distance"]
        assert len(network_long_df) == len(cspd_tp_df)

        # Read CARE/Overburdened mapping
        link_crosswalk_file = os.path.join(NETWORK_DIRS[network], f"link_to_COUNTY_{analysis_name}.csv")
        link_crosswalk = pd.read_csv(link_crosswalk_file)
        print(f"Read {len(link_crosswalk)} lines from {link_crosswalk}; head:")
        print(link_crosswalk.head())

        # Combine with network_long
        network_long_crosswalk_df = network_long_df.merge(
            link_crosswalk[["A", "B", "COUNTYCARE", "linkCC_share"]],
            left_on=["a", "b"], right_on=["A", "B"], how="left"
        )
        #TODO: filters
        network_long_crosswalk_df["county_census"] = network_long_crosswalk_df["COUNTYCARE"].str[:3]
        network_long_crosswalk_df["CARE"] = network_long_crosswalk_df["COUNTYCARE"].str.len() > 3

        # Filter out dummy links
        network_long_crosswalk_df = network_long_crosswalk_df[network_long_crosswalk_df["ft"] != 6]
        print(f"network_long_crosswalk_df has {len(network_long_crosswalk_df)} rows")

        # Read exhaust-based emissions
        if f"AL-{model_year}" not in emissions_rates:
            emissions_file = os.path.join(EMISSION_RATES_DIR, f"Year {model_year} MSAT Emission Rates with E2021 (PBA2050).xlsx")
            for _, row in LOOKUP_COUNTY.iterrows():
                county2 = row["county2"]
                sheet_name = f"{county2}-{model_year}.csv"
                emissions_df = pd.read_excel(emissions_file, sheet_name=sheet_name)
                emissions_df.columns = emissions_df.columns.str.replace("-", "_").str.replace(".", "_")
                emissions_rates[f"{county2}-{model_year}"] = emissions_df

        # Read non-exhaust-based emission rates
        non_exhaust_key = f"non-exhaust-{model_year}"
        if non_exhaust_key not in emissions_rates:
            emissions_file = os.path.join(EMISSION_RATES_DIR, "PM2.5 Non-Exhaust ERs (w E2021 & Mar 2021 road dust update).xlsx")
            emissions_df = pd.read_excel(emissions_file, sheet_name=model_year, skiprows=1)
            emissions_df.columns = emissions_df.columns.str.replace(" ", "_")
            emissions_rates[non_exhaust_key] = emissions_df

        # Initialize emissions columns
        network_long_crosswalk_df = network_long_crosswalk_df.assign(
            EO_PM2_5=0.0, EO_Benzene=0.0, EO_Butadiene=0.0, EO_DieselPM=0.0,
            Tire_Wear=0.0, Brake_Wear=0.0, Entrained_Road_Dust=0.0
        )

        # Process emissions for each county
        network_long_emissions_df = pd.DataFrame()

        for _, row in LOOKUP_COUNTY.iterrows():
            county2 = row["county2"]
            county_census = row["county_census"]
            county_name = row["county_name"]

            index_name = f"{county2}-{model_year}"
            non_exhaust_rates = emissions_rates[non_exhaust_key].query(f"County == '{county_name}'")

            # Filter for this county
            this_co_df = network_long_crosswalk_df.query(f"county_census == '{county_census}'")

            # Join based on congested speed int
            print(county_name)
            print('this_co_df')
            print(this_co_df.head())
            print(list(this_co_df))
            print('emissions_rates[index_name]')
            print(emissions_rates[index_name].head())
            print(list(emissions_rates))
            this_co_df = this_co_df.merge(emissions_rates[index_name], left_on="cspd_int", right_on="Speed", how="left")
            print('this_co_df after merge')
            print(this_co_df.head())
            this_co_df = this_co_df.assign(
                EO_PM2_5=np.where(this_co_df["CARE"], this_co_df["C_EO_PM2_5"], this_co_df["NC_EO_PM2_5"]),
                EO_Benzene=np.where(this_co_df["CARE"], this_co_df["C_EO_Benzene"], this_co_df["NC_EO_Benzene"]),
                EO_Butadiene=np.where(this_co_df["CARE"], this_co_df["C_EO_Butadiene"], this_co_df["NC_EO_Butadiene"]),
                EO_DieselPM=np.where(this_co_df["CARE"], this_co_df["C_EO_DieselPM"], this_co_df["NC_EO_DieselPM"]),
                Tire_Wear=np.where(this_co_df["CARE"], non_exhaust_rates["C_Tire_Wear"].values[0], non_exhaust_rates["NC_Tire_Wear"].values[0]),
                Brake_Wear=np.where(this_co_df["CARE"], non_exhaust_rates["C_Brake_Wear"].values[0], non_exhaust_rates["NC_Brake_Wear"].values[0]),
                Entrained_Road_Dust=np.where(this_co_df["CARE"], non_exhaust_rates["C_Entrained_Road_Dust"].values[0], non_exhaust_rates["NC_Entrained_Road_Dust"].values[0])
            )
            this_co_df = this_co_df.drop(columns=["C_EO_PM2_5", "NC_EO_PM2_5", "C_EO_Benzene", "NC_EO_Benzene",
                                                "C_EO_Butadiene", "NC_EO_Butadiene", "C_EO_DieselPM", "NC_EO_DieselPM"])

            print('this_co_df after assign')
            print(this_co_df.head())
            network_long_emissions_df = pd.concat([network_long_emissions_df, this_co_df])

        print(list(network_long_emissions_df))
        print(network_long_emissions_df)
        # Write the result
        if my_args.rtp == 'RTP2021':
            output_fullpath = os.path.join(BASE_DIR, "links_CARE_EMFAC2021_FEIR", f"links_CARE_{model_dir}.csv")
        elif my_args.rtp == 'RTP2025':
            output_fullpath = os.path.join(BASE_DIR, "links_OverBurdened_EMFAC2021_PBA50plus", f"links_OverBurdened_{model_dir}.csv")
        
        network_long_emissions_df.to_csv(output_fullpath, float_format='%.7f', index=False)
        print(f"Wrote {len(network_long_emissions_df)} rows and {len(list(network_long_emissions_df))} columns to {output_fullpath}")
